gemma.cc

```diff
@@ -88,89 +94,165 @@ struct Layer {
 
   std::array<float, TConfig::kModelDim> final_norm_scale;
+
+  LayerPointers<TConfig> layer_ptrs;
+
+  std::array<float, TConfig::kNumTensorScales> scales;
+
+  const Layer<TConfig>* GetLayer(size_t layer) const {
+    return layer_ptrs.layers[layer].get();
+  }
+  Layer<TConfig>* GetLayer(size_t layer) {
+    return layer_ptrs.layers[layer].get();
+  }
 };
 
-// Only called if cached loading fails.
 template <typename TConfig>
-hwy::AlignedUniquePtr<Weights<TConfig>> LoadWeights(const Path& checkpoint) {
+hwy::AlignedFreeUniquePtr<uint8_t[]> LoadWeights(
+    const Path& checkpoint, hwy::ThreadPool& pool,
+    bool scale_for_compression = false) {
   PROFILER_ZONE("Startup.LoadWeights");
-  using TWeights = Weights<TConfig>;
-  hwy::AlignedUniquePtr<TWeights> weights = hwy::MakeUniqueAligned<TWeights>();
-  weights->layers =
-      hwy::MakeUniqueAlignedArray<Layer<TConfig>>(TConfig::kLayers);
-
-  if (checkpoint.path.empty()) {
-    HWY_ABORT(
-        "Loading --compressed_weights failed; we require a --weights argument. "
-        "Please see issue #11 on how to create this file.\n");
+  if (!std::filesystem::exists(checkpoint.path)) {
+    HWY_ABORT("The model weights file '%s' does not exist.",
+              checkpoint.path.c_str());
   }
 
+  using TWeights = Weights<TConfig>;
+  hwy::AlignedFreeUniquePtr<uint8_t[]> weights_u8 =
+      hwy::AllocateAligned<uint8_t>(sizeof(TWeights));
+  TWeights* weights = reinterpret_cast<TWeights*>(weights_u8.get());
+  new (&weights->layer_ptrs) LayerPointers<TConfig>(pool);
+
+  size_t scale_pos = 0;
   FILE* fptr;
-  fptr = fopen(checkpoint.path.c_str(), "rb");
-  if (fptr == nullptr) {
-    HWY_ABORT("Failed to open model file %s - does it exist?",
-              checkpoint.path.c_str());
+  if constexpr (kDryRunFread) {
+    fprintf(stderr, "Dry-Run, not reading model-file.\n");
+  } else {
+    fptr = fopen(checkpoint.path.c_str(), "rb");
+    if (fptr == nullptr) {
+      HWY_ABORT("Failed to open model file %s - does it exist?",
+                checkpoint.path.c_str());
+    }
   }
   bool ok = true;
   uint64_t total_size = 0;
-  ok &= 1 == fread(&(weights->embedder_input_embedding),
-                   sizeof(weights->embedder_input_embedding), 1, fptr);
-  ok &= 1 == fread(&(weights->final_norm_scale),
-                   sizeof(weights->final_norm_scale), 1, fptr);
-  total_size += sizeof(weights->embedder_input_embedding) +
-                sizeof(weights->final_norm_scale);
+  auto do_fread = [&](void* var, size_t layer, const char* name, size_t size) {
+    if (layer == -1) {
+      fprintf(stderr, "Loading Parameters (size %zu): %s\n", size, name);
+    } else {
+      fprintf(stderr, "Loading Parameters (layer=%zu, size %zu): %s\n", layer,
+              size, name);
+    }
+    if constexpr (!kDryRunFread) {
+      ok &= 1 == fread(var, size, 1, fptr);
+      total_size += size;
+    }
+  };
+  do_fread(&(weights->embedder_input_embedding), -1, "embedder_input_embedding",
```

Comment:
-1 for size_t will at least raise warnings. Let's change do_fread arg to int, the layer=%zu to %d.

Reply:
Done.